language/vm/src/errors.rs

```diff
@@ -32,98 +32,299 @@ pub const EINSUFFICIENT_PRIVILEGES: u64 = 11; // user lacks the credentials to d
 
 pub const EASSERT_ERROR: u64 = 42; // catch-all error code for assert failures
 
-pub type VMResult<T> = ::std::result::Result<T, VMStatus>;
-pub type BinaryLoaderResult<T> = ::std::result::Result<T, VMStatus>;
+pub type VMResult<T> = ::std::result::Result<T, VMError>;
+pub type BinaryLoaderResult<T> = ::std::result::Result<T, PartialVMError>;
+pub type PartialVMResult<T> = ::std::result::Result<T, PartialVMError>;
+
+#[derive(Debug)]
+pub enum Location {
+    Unknown,
+    Script,
+    Module(ModuleId),
+}
+
+#[derive(Debug)]
+pub struct VMError {
+    pub major_status: StatusCode,
+    pub sub_status: Option<u64>,
+    pub message: Option<String>,
+    pub location: Location,
```

Comment:
Wondering what the location of an `EXECUTED` major_status would be. Would it be Script?

Reply:
It is most likely Unknown

Reply:
Maybe "Undefined" would be better than Unknown. Thoughts?

Reply:
hmmm, not sure what I would use here... I don't like either, though they are both fine.
Maybe `NoLocation`? :)

Reply:
I want something that really describes this thing comes from out in the ether, not inside a specific Move context. So I think NoLocation is good, but Undefined better captures this ethereal feeling. 

I feel like there is a better word out there, but I cannot find it

Reply:
after reading your summary in the slack channel I see what you are after.
`Internal`?

Reply:
Maybe... Sometimes it sort feels like its tied to a module, but it's not clear exactly whats going on. I'll ship it for now, and we should all get a better feeling